src/game/default/g_client_view.c

```diff
@@ -185,109 +185,106 @@ static void G_ClientWorldAngles(g_entity_t *ent) {
-
-	for (int32_t i = 0; i < 3; i++) {
-		if (kick[i] > 180.0) {
-			kick[i] -= 360.0;
-		}
-	}
-
-	// add in any event-based feedback
+static void G_ClientEventKick(g_entity_t *ent) {
 
 	switch (ent->s.event) {
 		case EV_CLIENT_LAND:
-			kick[PITCH] += 2.5;
+			G_ClientAddKick(ent, 2.5, 0.0);
 			break;
 		case EV_CLIENT_FALL:
-			kick[PITCH] += 5.0;
+			G_ClientAddKick(ent, 5.0, 0.0);
 			break;
 		case EV_CLIENT_FALL_FAR:
-			kick[PITCH] += 10.0;
+			G_ClientAddKick(ent, 10.0, 0.0);
 			break;
 		default:
 			break;
 	}
+}
 
-	// and any velocity-based feedback
+/**
+ * @brief Reduces client view kick angles over time, using exponential decay.
+ */
+static void G_ClientDecayKick(g_entity_t *ent) {
 
-	vec_t forward = DotProduct(ent->locals.velocity, ent->client->locals.forward);
-	kick[PITCH] += forward / 600.0;
+	vec_t delta = VectorLength(ent->client->locals.kick_angles);
+	delta = 0.250 + delta * delta * QUETOO_TICK_SECONDS;
 
-	vec_t right = DotProduct(ent->locals.velocity, ent->client->locals.right);
-	kick[ROLL] += right / 600.0;
+	for (int32_t i = 0; i < 3; i++) {
 
-	// now interpolate the kick angles towards neutral over time
+		// clear angles smaller than our delta to avoid oscillations
+		if (fabs(ent->client->locals.kick_angles[i]) <= delta) {
+			ent->client->locals.kick_angles[i] = 0.0;
+		} else if (ent->client->locals.kick_angles[i] > 0.0) {
+			ent->client->locals.kick_angles[i] -= delta;
+		} else {
+			ent->client->locals.kick_angles[i] += delta;
+		}
+	}
+}
+
+/**
+ * @brief Calculates the final view kick angles for the current frame. This factors in damage, 
+ * weapon and event kicks, and uses interpolation and exponential decay to produce a smooth effect.
+ */
+static void G_ClientKickAngles(g_entity_t *ent) {
+	uint16_t *kick_angles = ent->client->ps.pm_state.kick_angles;
 
-	vec_t delta = VectorLength(kick);
+	// spectators and dead clients receive no kick angles
 
-	if (!delta) { // no kick, we're done
+	if (ent->client->ps.pm_state.type != PM_NORMAL) {
+		VectorClear(kick_angles);
 		return;
 	}
 
-	// we recover from kick at a rate based on the kick itself
+	G_ClientEventKick(ent);
 
-	delta = 0.5 + delta * delta * QUETOO_TICK_SECONDS;
+	vec3_t kick;
+	UnpackAngles(kick_angles, kick);
 
-	for (int32_t i = 0; i < 3; i++) {
+	// if a kick has recently occurred, interpolate it over a few frames
 
-		// clear angles smaller than our delta to avoid oscillations
-		if (fabs(kick[i]) <= delta) {
-			kick[i] = 0.0;
-		} else if (kick[i] > 0.0) {
-			kick[i] -= delta;
-		} else {
-			kick[i] += delta;
-		}
+	const vec_t frac = (g_level.time - ent->client->locals.kick_angles_time) / 64.0;
+	if (frac < 1.0) {
+		AngleLerp(kick, ent->client->locals.kick_angles, Clamp(frac, 0.0, 1.0), kick);
+	} else {
+		// otherwise, decay the kick, and use the resulting value directly
+		G_ClientDecayKick(ent);
+		VectorCopy(ent->client->locals.kick_angles, kick);
 	}
 
 	PackAngles(kick, kick_angles);
```

Comment:
Well this is fucking weird. Why is this showing up here? I pushed this to `master` already.